api/server/sdk/credentials.go

```diff
@@ -0,0 +1,383 @@
+	// respective cloud provider and return result
+	// this may not be expected behaviour, we have to do this since
+	// `interface` can't be mapped directly with other lang
+	azureCreds, err := getCredentialMap(credList, "azure")
+	if err != nil {
+		return nil, status.Errorf(
+			codes.Internal,
+			"Unable to enumerate credentials Azure: %v",
+			err.Error())
+	}
+
+	// Fill up azure credential resonse
+	creds := []*api.AzureCredential{}
+	for k, v := range azureCreds {
+		cred, ok := v.(map[string]interface{})
+		if !ok {
+			return nil, status.Errorf(
+				codes.Internal,
+				"Unable to enumerate credentials AWS: %v",
+				reflect.TypeOf(v).String())
+		}
+
+		credResp := &api.AzureCredential{
+			CredentialId: k,
+			AccountName:  cred[api.OptCredAzureAccountName].(string),
+			AccountKey:   cred[api.OptCredAzureAccountKey].(string),
+		}
+		creds = append(creds, credResp)
+	}
+	return &api.CredentialEnumerateAzureResponse{Credential: creds}, nil
+}
+
+// EnumerateForGoogle list credentials for Google
+func (s *VolumeServer) EnumerateForGoogle(
+	ctx context.Context,
+	req *api.CredentialEnumerateGoogleRequest,
+) (*api.CredentialEnumerateGoogleResponse, error) {
+	credList, err := s.driver.CredsEnumerate()
+	if err != nil {
+		return nil, status.Errorf(
+			codes.Internal,
+			"Unable to enumerate credentials: %v",
+			err.Error())
+	}
+
+	// By defaultcredList will have all credential details, we will extract for
+	// respective cloud provider and return result
+	// this may not be expected behaviour, we have to do this since
+	// `interface` can't be mapped directly with other lang
+	googleCreds, err := getCredentialMap(credList, "google")
+	if err != nil {
+		return nil, status.Errorf(
+			codes.Internal,
+			"Unable to enumerate credentials Azure: %v",
+			err.Error())
+	}
+
+	// Fill up google credential resonse
+	creds := []*api.GoogleCredential{}
+	for k, v := range googleCreds {
+		cred, ok := v.(map[string]interface{})
+		if !ok {
+			return nil, status.Errorf(
+				codes.Internal,
+				"Unable to enumerate credentials AWS: %v",
+				reflect.TypeOf(v).String())
+		}
+
+		credResp := &api.GoogleCredential{
+			CredentialId: k,
+			ProjectId:    cred[api.OptCredGoogleProjectID].(string),
+		}
+		creds = append(creds, credResp)
+	}
+
+	return &api.CredentialEnumerateGoogleResponse{Credential: creds}, nil
+}
+
+func validateAndDelete(s *VolumeServer, uuid string) error {
+	// Validate if the credentials provided were correct or not
+	req := &api.CredentialValidateRequest{CredentialId: uuid}
+
+	validateErr := s.driver.CredsValidate(req.GetCredentialId())
+
+	if validateErr != nil {
+		deleteCred := &api.CredentialDeleteRequest{CredentialId: uuid}
+		err := s.driver.CredsDelete(deleteCred.GetCredentialId())
+
+		if err != nil {
+			return status.Errorf(
+				codes.Internal,
+				"failed to delete invalid Google credentials: %v",
+				err.Error())
+		}
+
+		return status.Errorf(
+			codes.Internal,
```

Comment:
If the provided credentials did not validate, we shouldn't use `codes.Internal` as an error. I think we should use [PermissionDenied](https://github.com/grpc/grpc-go/blob/master/codes/codes.go#L72).